### ISSUE TITLE
Fail to declare a typed.Dict inside numba.jit

### ISSUE THREAD

**Issue by BenjaminChoou** [numba/numba#7301]
## Reporting a bug

- [Yes] I have tried using the latest released version of Numba (most recent is
 visible in the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [Yes] I have included a self contained code sample to reproduce the problem.
  i.e. it's possible to run as 'python bug.py'.

I want to create a dict inside jit function but gets
```
numba.core.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Unknown attribute 'Tuple' of type Module(<module 'numba.core.types' from '/Users/benjaminchou/miniforge3/envs/tf2/lib/python3.8/site-packages/numba/core/types/__init__.py'>)
```

The code to reproduce is:
```
import numba
from numba import objmode, typed, types


@numba.jit(nopython=True)
def dict_test():
    d = typed.Dict.empty(
        key_type=types.Tuple((types.int64, types.int64)),
        value_type=types.int64,
    )
    elem = (1, 2)
    d[elem] = d.get(elem, 0) + 1
    return d
```

A workaround is:
```
import numba
from numba import objmode, typed, types


@numba.jit(nopython=True)
def dict_test():
    d = typed.Dict.empty(
        key_type=tuple((1, 1)),
        value_type=types.int64,
    )
    elem = (1, 2)
    d[elem] = d.get(elem, 0) + 1
    return d
```

Could anyone explain the correct way to use `typed.Dict`? So far, I don't find any detailed example in the docs.

**Comment by sklam** [comment]
Thanks you for reporting the issue. The dictionary support can be unintuitive sometimes because of limitations from type inference. 

It does not work in the reproducer because Numba cannot handle the type spelling (`types.Tuple((types.int64, types.int64))`) in jit code. If you move that into a global, it will work. For example:

```python
import numba
from numba import typed, types

key_type = types.Tuple((types.int64, types.int64))

@numba.jit(nopython=True)
def dict_test():
    d = typed.Dict.empty(
        key_type=key_type,  # this now references a global
        value_type=types.int64,
    )
    elem = (1, 2)
    d[elem] = d.get(elem, 0) + 1
    return d

```

In simpler cases, Numba can infer the type of the `dict` so you can spell it with the typical `d = {}`. But, your case here is a little bit challenging for type inference because `d.get(elem, 0)` is used before the `d[elem] = ...`. 

For example, the following case can be type inferred:

```python
@numba.jit(nopython=True)
def dict_test():
    elem = (1, 2)
    d = {elem: 0}
    d[elem] = d.get(elem, 0) + 1
    return d
```

Lastly, I'm a bit surprised that the workaround spelling is accepted. 

```python
d = typed.Dict.empty(
        key_type=tuple((1, 1)),  # surprised this is accepted
        value_type=types.int64,
    )
```

**Comment by BenjaminChoou** [comment]
> Thanks you for reporting the issue. The dictionary support can be unintuitive sometimes because of limitations from type inference.
> 
> It does not work in the reproducer because Numba cannot handle the type spelling (`types.Tuple((types.int64, types.int64))`) in jit code. If you move that into a global, it will work. For example:
> 
> ```python
> import numba
> from numba import typed, types
> 
> key_type = types.Tuple((types.int64, types.int64))
> 
> @numba.jit(nopython=True)
> def dict_test():
>     d = typed.Dict.empty(
>         key_type=key_type,  # this now references a global
>         value_type=types.int64,
>     )
>     elem = (1, 2)
>     d[elem] = d.get(elem, 0) + 1
>     return d
> ```
> 
> In simpler cases, Numba can infer the type of the `dict` so you can spell it with the typical `d = {}`. But, your case here is a little bit challenging for type inference because `d.get(elem, 0)` is used before the `d[elem] = ...`.
> 
> For example, the following case can be type inferred:
> 
> ```python
> @numba.jit(nopython=True)
> def dict_test():
>     elem = (1, 2)
>     d = {elem: 0}
>     d[elem] = d.get(elem, 0) + 1
>     return d
> ```
> 
> Lastly, I'm a bit surprised that the workaround spelling is accepted.
> 
> ```python
> d = typed.Dict.empty(
>         key_type=tuple((1, 1)),  # surprised this is accepted
>         value_type=types.int64,
>     )
> ```

Yes. I just find that python instances can be inferred for types in the jitted function. That's something not in the docs, which makes me quite confused. I am kind of lost in the usage of signatures. It's not a very consistent experience. As for your workaround moving it to global, it requires me to manage a global space for signatures, which is quite oblique. 
I guess it's because most codes of `types` work in python-mode while not supported in non-python mode. I wonder if it is possible to provide `types` argument in `jit` to mimic the C++ template?

**Comment by sklam** [comment]
Yes, you can also pass types as parameters:

e.g.
```python
import numba
from numba import typed, types


@numba.jit(nopython=True)
def dict_test(key_type):
    d = typed.Dict.empty(
        key_type=key_type,  # this is now from an arg
        value_type=types.int64,
    )
    elem = (1, 2)
    d[elem] = d.get(elem, 0) + 1
    return d

tup_type = types.Tuple((types.int64, types.int64))
r = dict_test(tup_type)
print(r)
```

**Comment by esc** [comment]
Closing this now as the question has been answered. Please feel free to re-open in case the issue remains.
